constantine/boolean_hypercube/multilinear_extensions.nim

```diff
@@ -0,0 +1,318 @@
+# Constantine
+# Copyright (c) 2018-2019    Status Research & Development GmbH
+# Copyright (c) 2020-Present Mamy André-Ratsimbazafy
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+
+import
+  constantine/platforms/[abstractions, allocs]
+
+# Resources:
+# - https://people.cs.georgetown.edu/jthaler/IPsandextensions.pdf
+# - https://people.cs.georgetown.edu/jthaler/ProofsArgsAndZK.pdf
+#   Chapter 3.5
+
+type
+  MultilinearExtension*[F] = object
+    ## Note: to follow mathematical description, indices start at 1 and end of range is inclusive
+    ##       actual implementation will have indices start at 0
+    ##
+    ## Given a sequence of bits of size s {0,1}ˢ
+    ## and an arbitrary function f: {0,1}ˢ -> 𝔽
+    ## i.e. that maps a sequence of bits to a finite field 𝔽
+    ##
+    ## there is an unique multilinear polynomial f̃
+    ## called multilinear extension of f
+    ## that satisfies f̃(xᵢ) = f(xᵢ) for xᵢ ∈ {0,1}ˢ
+    ##
+    ## f̃(x₁, ...,xₛ) = ∑ₑ f(e) 𝛘ₑ(x₁, ...,xₛ)
+    ## with e ∈ {0,1}ˢ and f(e) the evaluation of f at e.
+    ##
+    ## 𝛘ₑ(x₁, ...,xₛ) is the multilinear Lagrange basis polynomial
+    ## which takes value 1 at 𝛘ₑ(e) and 0 at Xₑ(e̅) e̅ any other element ∈ {0,1}ˢ
+    ##
+    ## 𝛘ₑ(x₁, ...,xₛ) = ∏ᵢ₌₁ˢ(xᵢeᵢ + (1-xᵢ)(1-eᵢ)), i ∈ [1, s]
+    ##
+    ## A multilinear polynomial is linear (i.e. degree at most 1) in each
+    ## of its variables.
+    num_vars*: int
+    base_poly_evals*: ptr UncheckedArray[F]
+
+proc `=destroy`*[F](mle: var MultilinearExtension[F]) =
```

Comment:
If we attach a destructor to a plain (non ref) `object`, we should better error on `=copy`, no? Otherwise there is a risk of a copy of a `MultilinearExtension` object going out of scope and calling the destructor and thus freeing the memory. 

(my experience with `object` + `=destroy` has been less than splendid).

Reply:
Right, that or check if nil before calling free.

Reply:
No, I mean not in terms of a potential double free, but in terms of having a dangling pointer in the instance that outlives the other. One might accidentally free the storage too early.

Reply:
Addressed in https://github.com/mratsim/constantine/pull/423/commits/acbfc482eb78fa762406c2a3f86e4fa17a5b112b